File: services/userService.js
import { userRepository } from "@/repositories/userRepository";
import bcrypt from "bcrypt";


export const userService = {

    async getUserId(usernameReq) {
        if(typeof usernameReq === "string") {
            try {
                const userId = await userRepository.getUserId(usernameReq);
                return userId?.id;
            } catch(error) {
                console.error("Service layer error for getting user id: " + error);
                return null;
            }
        }
    },

    async register(user) {

        try {
            const userId = await this.getUserId(user.username);
            if(userId != null) {
                throw new Error("The username is already taken");
            } else {
                const hashedPassword = await this.hashPassword(user.password);
                user.password = hashedPassword;
                return await userRepository.addUser(user);
            }

        } catch(error) {
            console.error("Service layer error for registration: " + error)
            throw error;
        }

    },

    async login(user) {
        try {
            const storedUser = await userRepository.getUserByUsername(user.username);
            
            if (!storedUser) {
                throw new Error("Wrong username.");
            }

            const passwordMatches = await this.passwordMatches(user.password,storedUser.password);

            if (!passwordMatches) {
                throw new Error("Wrong password.");
            }

            return {
                id: storedUser.id,
                username: storedUser.username,
                message: "Login successful",
            };
        } catch (error) {
            console.error("Service layer error for login: " + error.message);
            throw error;
        }
    },

    async logout(user) {
        const userId = await this.getUserId(user.username);
        
        if (!userId) {
            throw new Error("User not found.");
        }

        await userRepository.logout(userId);

        return {
            id: userId,
            username: user.username,
            message: "Logout successful",
        };
        
    },

    async passwordMatches(plainPassword, hashedPassword) {
        return bcrypt.compare(plainPassword, hashedPassword);
    },

    async hashPassword(password) {
        const saltRounds = 10;

        const hashedPassword = await bcrypt.hash(password, saltRounds);

        return hashedPassword;
    }
}

File: services/user_service.rb
class UserService
    def self.get_user_id(username)
        user = User.find_by(username: username)
        user&.index
    rescue => e
        Rails.logger.error("Service error getting user ID: #{e}")
        nil
    end

    def self.register(params)
        if get_user_id(params[:username])
            raise "Username already exists"
        end

        user = User.new(
            username: params[:username],
            password: params[:password]
        )

        if user.save
            user
        else
            raise StandardError.new(user.errors.full_messages.join(", "))
        end
    end

    def self.login(username:, password:)
        user = User.find_by(username: username)
        raise StandardError.new("Invalid username") unless user

        unless user.authenticate(password)
            raise StandardError.new("Invalid password")
        end

                # Update last_login timestamp
                begin
                    user.update_columns(last_login: Time.current)
                rescue => e
                    Rails.logger.error("Failed to update last_login for user=#{user.id}: #{e}")
                end

                { id: user.id, username: user.username, message: "Login successful" }
    end

    def self.logout(username)
        user = User.find_by(username: username)
        raise StandardError.new("User not found") unless user

        { id: user.id, username: user.username, message: "Logout successful" }
    end
end
